os/fs/smartfs/smartfs_smart.c

```diff
@@ -289,21 +288,40 @@ static int smartfs_open(FAR struct file *filep, const char *relpath, int oflags,
 			}
 
 			/* First try to find an invalid or empty entry available in one of the chained parent sectors */
-			ret = smartfs_find_availableentry(fs, &parentdirsector, &newentry);
+			ret = smartfs_find_availableentry(fs, &parentdirsector, &sf->entry);
 			if (ret != OK) {
 				/* find_availableentry encountered a problem with lower layer operations, return error */
 				fdbg("smartfs_find_availableentry() encountered a problem, unable to find entry for writing\n");
 				goto errout_with_buffer;
 			}
 
 			/* At this point, either an available entry was found or a new one has been created */
-			newentry.firstsector = data_sector;
-
-			ret = smartfs_writeentry(fs, newentry, filename, SMARTFS_DIRENT_TYPE_FILE, mode, &sf->entry, parentdirsector);
+			sf->entry.firstsector = data_sector;
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+			/* If CRC is enabled, hold the entry for writing later */
+			/* Mark flags to indicate that the entry will be written to anew sector in the parent directory.
+			   This sector will then be chained in the end.
+			 */
+			sf->last_psector = parentdirsector;
+			sf->bflags |= SMARTFS_BFLAG_NEW_ENTRY;
+			sf->mode = mode;
+			if (sf->entry.name == NULL) {
```

Comment:
hmm ok, in next pr please think how we can reduce duplicated code here =)
(here & writeentry)
Then regardless of crc, we can use it.

Reply:
Yes, I want to avoid so many code lines for allocating space for name.
Let's try for a macro/function in the future.

Reply:
I mean I think this will be set 2times, here and writeentry.
Instead of using the MACRO, we'd better remove it from somewhere.

Reply:
Here, allocating space for entry name is  necessary to store the entry name for writing later.
If the space is allocated here, it will not be allocated in writeentry for direntry.

However, before line 1334 of smartfs_utils.c
I think space needs to be allocated for entry->name.
We might have missed out on that...

Reply:
It is needed only to create new entry, so If we do this in smartfs_find_availableentry, we can reduce it. please consider it.
Also smartfs_createentry is needed to create new entry sector when it couldn't find entry parent directory sector , so rename is needed smartfs_createdirentry (I think I already reviewed it).

Reply:
smartfs_find_availableentry only finds the space for writing new entry.
Allocating space for new entry->name is done while writing only.

So this should not be done in find_availableentry.

`filename `is copied to RAM buffer only when we are populating attributes of entry to be written.
Before that let's not allocate entry->name separately.

Reply:
I don't think so, 'creating entry' means that find space & make entry struct.
Entry can be hold both of flash or RAM. Even if we store it in RAM, it already completed entry.
So you can do something more here =)
Please do not restrict to use current code only.
You can rename current function & expand it, or add more function.
Current writeentry doing 2 roles, update entry info and write entry data on flash.
How to optimize below?
```
main()
{
    func_a();
    func_b();
    if (condition == true) {
        func_c(); // function_c set data only
        func_d();
    } else {
        func_e();
    }
    ...
    if (condition == true) {
        func_e();
    }
}

func_e()
{
    func_c();
    func_f();
}
```
Expand or rename function a or b, otherwise add new function g, which can call func_a, func_b, func_c will be one of answer.